tests/functional/test_snap.py

```diff
@@ -1,5 +1,157 @@
+SNAP_NAME = "smartctl-exporter"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_service_failed() -> None:
+    """Check if a systemd service is in a failed state."""
+    assert 0 == subprocess.call(
+        f"sudo systemctl is-failed --quiet {SYSTEMD_SERVICE}".split()
+    ), f"{SYSTEMD_SERVICE} is running"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_service_active() -> None:
+    """Check if a systemd service is in a activemstate."""
+    assert 0 == subprocess.call(
+        f"sudo systemctl is-active --quiet {SYSTEMD_SERVICE}".split()
+    ), f"{SYSTEMD_SERVICE} is not running"
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def _check_endpoint(endpoint: str) -> None:
+    """Check if an endpoint is reachable."""
+    response = urllib.request.urlopen(endpoint)  # will raise if not reachable
+    status_code = response.getcode()
+    assert status_code == 200, f"Endpoint {endpoint} returned status code {status_code}"
+
+
+@retry(wait=wait_fixed(5), stop=stop_after_delay(30))
+def _get_endpoint_data(endpoint: str) -> str:
+    """Get the data from an endpoint."""
+    response = urllib.request.urlopen(endpoint)
+    return response.read().decode("utf-8")
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _check_bind(bind: str, service: str) -> None:
+    """Check if a service is listening on a specific bind."""
+    pid = subprocess.check_output(f"sudo lsof -t -i {bind}".split(), text=True).strip()
+    assert service in subprocess.check_output(
+        f"cat /proc/{pid}/cmdline".split(), text=True
+    ), f"{SNAP_NAME} is not listening on {bind}"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _set_config(config_parent: str, config: str, value: str) -> None:
+    """Set a configuration value for a snap service."""
+    assert 0 == subprocess.call(
+        f"sudo snap set {SNAP_NAME} {config_parent}.{config}={value}".split()
+    ), f"Failed to set {config_parent}.{config} to {value}"
+
+
+@retry(wait=wait_fixed(2), stop=stop_after_delay(10))
+def _unset_config(config_parent: str, config: str) -> None:
+    """Unset a configuration value for a snap service."""
+    assert 0 == subprocess.call(
+        f"sudo snap unset {SNAP_NAME} {config_parent}.{config}".split()
+    ), f"Failed to unset {config}"
+
+
+def _get_start_cmd(process: str) -> str:
+    return subprocess.check_output(f"ps -C {process} -o cmd -ww".split(), text=True)
+
+
+def _check_config(config_parent: str, config: str):
+    """Check if a configuration exists in the snap configuration."""
+    result = subprocess.check_output(f"sudo snap get {SNAP_NAME} -d".split(), text=True)
+    smartctl_snap_config = json.loads(result.strip())
+    assert (
+        config in smartctl_snap_config[config_parent]
+    ), f"{config} is not in the snap configuration"
+
+
+@contextmanager
+def _config_setup(config_str, new_value):
+    """Set up a context manager to test snap configuration."""
```

Comment:
This could have better docs and naming, to include that it's a context manager so you can use it to test the snap with custom config, and have the config reverted afterwards even if the test fails.

I like this pattern - it makes the tests really clear and concise. :+1: